[Feat] 이미지 삭제 기능 구현 및 저장 로직 리팩토링

ghrltjdtprbs opened the pull request:
# ⭐️ [Feat] 이미지 삭제 기능 구현 및 저장 로직 리팩토링

- 이미지 삭제 기능을 구현하고 저장 로직을 리팩토링하였습니다.

## 🛠️ 변경 사항

### 유지보수 게시글 수정 dto에 이미지 제거
 - 이미지 등록 방식이 변경되어 유지보수 게시글에 이미지 수정 부분이 제거되어야함
- UpdateRepairArticleRequestDto 에 images 제거

### 게시글 저장시 이미지 id 리스트를 받도록 변경
 - 이미지 업로드 로직이 변경되어 기존의 이미지 url리스트가 아닌 이미지 id 리스트를 받아야합니다.
 - id 리스트를 받도록 관련 로직 수정


### SchedulerConfig 삭제
 - SchedulerConfig가 앞으로 사용 가능성이 없다고 판단.
 - SchedulerConfig클래스 삭제
 - Final10aeatApplication에 @EnableScheduling 추가

### 이미지 삭제기능 구현
 - 이미지 삭제 기능 구현
   DELETE /managers/repair/articles/image/{imageId}
 - 비정상적인 이미지 데이터 값과 이미 삭제된 게시글의 이미지들을 스케줄러로 일괄 삭제하는 로직 추가

### 이미지 업로드, 삭제 REST Docs 문서화

## 💡 관련 이슈

- #183 
- #185 

## 📃 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [x] 리팩토링
- [x] 테스트 코드 작성
- [ ] 문서 업데이트

## ⏱️ 작업 기간

- 작업 시작일: (2024-06-11)
- 작업 종료일: (2024-06-12)

## 📌 유의사항

- hotfix 브랜치로 잘못 보냈습니다....

ghrltjdtprbs committed:
[refactor] 서비스 계층에서의 권한 검사 중복 제거

변경 사항 :
- 이미지 업로드 관련 책임 분리를 위해 서비스 분리
- 이미 컨트롤러에서 @PreAuthorize 어노테이션을 통해 권한 검사를 하고 있으므로 saveImage에서의 권한 검사는 중복임 -> 제거

ghrltjdtprbs committed:
[refactor] 유지보수 게시글 수정 dto에 이미지 제거

문제 사항 :
 - 이미지 등록 방식이 변경되어 유지보수 게시글에 이미지 수정 부분이 제거되어야함

변경 사항 :
- UpdateRepairArticleRequestDto 에 images 제거

ghrltjdtprbs committed:
[feat] ImageNotFoundException 예외 클래스 추가

변경 사항 :
 - ImageNotFoundException 클래스 추가
 - ErrorCode 관련 메시지 추가

ghrltjdtprbs committed:
[refactor] 게시글 저장시 이미지 id 리스트를 받도록 변경

문제 사항 :
 - 이미지 업로드 로직이 변경되어 기존의 이미지 url리스트가 아닌 이미지 id 리스트를 받아야합니다.

변경 사항 :
 - id 리스트를 받도록 관련 로직 수정

ghrltjdtprbs committed:
[remove] SchedulerConfig 삭제

문제 사항 :
 - SchedulerConfig가 앞으로 사용 가능성이 없다고 판단.

변경 사항 :
 - SchedulerConfig클래스 삭제
 - Final10aeatApplication에 @EnableScheduling 추가

ghrltjdtprbs committed:
[feat] 이미지 삭제기능 구현

변경 사항 :
 - 이미지 삭제 기능 구현
   DELETE /managers/repair/articles/image/{imageId}
 - 비정상적인 이미지 데이터 값과 이미 삭제된 게시글의 이미지들을 스케줄러로 일괄 삭제하는 로직 추가

Galmaeki commented on src/main/java/com/final_10aeat/common/scheduler/DeleteArticleImageScheduler.java:
```diff
@@ -0,0 +1,22 @@
+    public void cleanupOrphanAndNonExistentImages() {
+        List<Long> imageIdsToDelete = repairArticleImageRepository.findIdsOfOrphanAndNonExistentRepairArticleImages();
+        if (!imageIdsToDelete.isEmpty()) {
+            repairArticleImageRepository.deleteAllByIdIn(imageIdsToDelete);
```
여기서 s3 삭제코드도 필요할것같습니다!

ghrltjdtprbs replied:
게시글 생성시 비정상적인 요청으로 인해 생성된 데이터라고 생각해 일부러 s3 삭제는 제외시켰는데, 다시 생각해보니 필요한 것 같습니다! 지적 감사합니다 ㅎㅎ

Galmaeki reviewed (changes requested):
코멘트 확인 부탁드림다

ghrltjdtprbs committed:
[refactor] RepairArticle의 image를 List로 변경

문제 사항 :
 - 게시글 이미지는 중복 상관 x
 - 순서를 보장해야함

변경 사항 :
 - RepairArticle의 image를 Set -> List 로 변경
 - 관련 로직 변경

ghrltjdtprbs committed:
[fix] 실패하는 테스트 코드 수정

변경 사항 :
 - 리팩토링으로 인해 실패하는 테스트 코드 수정

ghrltjdtprbs committed:
[test] 이미지 업로드, 삭제 REST Docs 문서화

변경 사항 :
 - 이미지 업로드, 삭제 기능 REST Docs 테스트 코드 작성
 - 문서화

ghrltjdtprbs committed:
[refactor] 스케줄러로 비정상적인 이미지 삭제시 s3에서도 삭제되도록 변경

문제 사항 :
 - 스케줄러가 비정상적인 이미지를 데이터베이스에서 삭제할뿐만 아니라 s3에서도 해당 이미지를 삭제하여야함

변경 사항 :
- DeleteArticleImageScheduler를 추가하여 매일 새벽 2시에 orphan 및 non-existent 이미지를 삭제하도록 설정
- RepairArticleImageRepository에 orphan 및 non-existent 이미지를 찾는 쿼리 추가
- ManagerRepairArticleImageService에 이미지 저장 및 삭제 시 S3에서도 처리하도록 로직 추가
- 필요 없는 이미지를 데이터베이스와 S3 버킷에서 모두 삭제하여 데이터 정합성 유지

Galmaeki reviewed (approved):
고생하셨습니다!

asd42270 reviewed (approved):
확인했습니다!! 고생하셨습니다!
